### PR TITLE
sdk/go: Add pulumix subpackage

### LINKED ISSUE #13586: [Go Generics] Implement the pulumix [Phase 1]
Per RFC #13057, the pulumix package will contain the core, type-safe Input/Output types, combinators, and generic containers built on top of them.

All existing types in sdk/go/pulumi will implement pulumi.Input[T] for their respective types.
```[tasklist]
### Tasks
- [ ] pulumix: Add Input, Output, Apply, etc.
- [ ] pulumix: Add Apply2, Apply3, ..., Apply8 and other combinators
- [ ] pulumi: Implement pulumix.Input[T] for all built-ins
- [ ] pulumix: Add Ptr, Array, and Map containers
```

### PR DESCRIPTION
**Pull request by abhinav**
**Each commit is separately reviewable.**

---

Adds the pulumix subpackage implementing strongly-typed
inputs, outputs, and combinators.
internal/cmd/gen-pux-applyn generates all four variants of `Apply` for
up to 8 parameters.

See individual commits for more details on each change.

Resolves #13586

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13495** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13495" target="_blank">Graphite</a> 
    * **PR #13509** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13509" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/13509?utm_source=stack-comment).

**Review by justinvp — APPROVED** [review]
LGTM

@Zaid-Ajaj, @Frassle did you also want to take a look?

**Review by Zaid-Ajaj — APPROVED** [review]

File: sdk/go/pulumix/types_ext_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	o := pulumix.Val("foo").AsAny()
+
+	v, known, secret, _, err := internal.AwaitOutput(context.Background(), o)
+	require.NoError(t, err)
+	assert.True(t, known)
+	assert.False(t, secret)
+	assert.Equal(t, "foo", v)
+}
+
+func TestOutput_MustCast(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (abhinav) [reply sdk/go/pulumix/types_ext_test.go:60]: I'll defer this to the upcoming rename.

**Comment by abhinav** [comment]
Canceled merge because 3.79 isn't fully released yet. This shouldn't accidentally become part of the release.

**Comment by Frassle** [comment]
> @Zaid-Ajaj, @Frassle did you also want to take a look?

Looks all reasonable to me

**Comment by FollieHiyuki** [comment]
Hi @abhinav! There seems to be an incorrect import line here: https://github.com/pulumi/pulumi/blob/362cc2a41acfe8b8b92185568e0ad9ab7ccf8d96/sdk/go/pulumi/generate/templates/types_builtins.go.template#L26

This module doesn't exist, so running `go generate` on this template will fail:
```go
../pulumi/types_builtins.go:26:2: no required module provides package github.com/pulumi/pulumi/sdk/v3/go/pulumi/internal; to add it:
	go get github.com/pulumi/pulumi/sdk/v3/go/pulumi/internal
```

The generated result that is committed into the repo doesn't have the import line: https://github.com/pulumi/pulumi/blob/362cc2a41acfe8b8b92185568e0ad9ab7ccf8d96/sdk/go/pulumi/types_builtins.go#L23-L27

**Comment by justinvp** [comment]
Thanks for the heads-up @FollieHiyuki! Opened https://github.com/pulumi/pulumi/issues/13871. I'm preparing a fix now.

**Pull request merged**